Bottom: a4d63da39044ba22a9f221d0f68c3c14b395471f
Top:    b3626f9279513682eaea584997a80902d06e513f
Author: Ian Kent <raven@themaw.net>
Date:   2013-06-07 17:16:46 +0800

debug.patch


---

diff --git a/lib/cache.c b/lib/cache.c
index 1e05a99..ecace4a 100644
--- a/lib/cache.c
+++ b/lib/cache.c
@@ -659,7 +659,7 @@ int cache_update_offset(struct mapent_cache *mc, const char *mkey, const char *k
 
 	me = cache_lookup_distinct(mc, key);
 	if (me && me->age == age) {
-		if (me->multi == owner)
+		if (me == owner || strcmp(me->key, key) == 0)
 			return CHE_DUPLICATE;
 	}
