Bottom: 4b9c00872f04d4b6f88d4006f53a90777eafcc33
Top:    662a846c8120f9bd9b0ff20f430fa9ca3781869e
Author: Ian Kent <ikent@redhat.com>
Date:   2013-09-23 14:41:26 +0800

debug.patch


---

diff --git a/lib/rpc_subs.c b/lib/rpc_subs.c
index 25f8968..a3444c4 100644
--- a/lib/rpc_subs.c
+++ b/lib/rpc_subs.c
@@ -588,8 +588,9 @@ int rpc_portmap_getclient(struct conn_info *info,
 	info->addr = addr;
 	info->addr_len = addr_len;
 	info->program = rpc_getrpcbyname(rpcb_prog);
-	info->port = rpc_getrpcbport(proto);
+	info->port = ntohs(rpc_getrpcbport(proto));
 	info->version = rpcb_version;
+	error(LOGOPT_ANY, "program %d port %d version %d", info->program, info->port, info->version);
 	info->proto = proto;
 	info->send_sz = RPCSMALLMSGSIZE;
 	info->recv_sz = RPCSMALLMSGSIZE;
@@ -632,9 +633,10 @@ int rpc_portmap_getport(struct conn_info *info,
 		pmap_info.host = info->host;
 		pmap_info.addr = info->addr;
 		pmap_info.addr_len = info->addr_len;
-		pmap_info.port = rpc_getrpcbport(info->proto);
+		pmap_info.port = ntohs(rpc_getrpcbport(info->proto));
 		pmap_info.program = rpc_getrpcbyname(rpcb_prog);
 		pmap_info.version = rpcb_version;
+		error(LOGOPT_ANY, "program %d port %d version %d", pmap_info.program, pmap_info.port, pmap_info.version);
 		pmap_info.proto = info->proto;
 		pmap_info.send_sz = RPCSMALLMSGSIZE;
 		pmap_info.recv_sz = RPCSMALLMSGSIZE;
