Bottom: a4d63da39044ba22a9f221d0f68c3c14b395471f
Top:    d1354e7b9d1b968ca9b674b3fd02b5d2f914ee94
Author: Ian Kent <raven@themaw.net>
Date:   2013-06-07 17:16:46 +0800

debug.patch


---

diff --git a/lib/cache.c b/lib/cache.c
index 1e05a99..7a2caa6 100644
--- a/lib/cache.c
+++ b/lib/cache.c
@@ -659,7 +659,7 @@ int cache_update_offset(struct mapent_cache *mc, const char *mkey, const char *k
 
 	me = cache_lookup_distinct(mc, key);
 	if (me && me->age == age) {
-		if (me->multi == owner)
+		if (me == owner)
 			return CHE_DUPLICATE;
 	}
