Bottom: 77e6f9a64fe21d866c09b5e3cf02a3b39e104bc2
Top:    1f9abd819fc97b7317f79364a7e4f3aae95677e0
Author: Ian Kent <ikent@redhat.com>
Date:   2013-04-08 09:57:09 +0800



---

diff --git a/lib/master.c b/lib/master.c
index a0e62f2..d12e672 100644
--- a/lib/master.c
+++ b/lib/master.c
@@ -907,7 +907,10 @@ int master_notify_submount(struct autofs_point *ap, const char *path, enum state
 
 		if (!master_submount_list_empty(this)) {
 			mounts_mutex_unlock(ap);
-			return master_notify_submount(this, path, state);
+			error(LOGOPT_ANY, "notify submount %s", this->path);
+			master_notify_submount(this, path, state);
+			mounts_mutex_lock(ap);
+			continue;
 		}
 
 		/* path not the same */
